### PR TITLE
Add isNonPositive and isNonNegative

### LINKED ISSUE #324: RA.isNonNegative, RA.isNonPositive
There are currently `isPositive` and `isNegative`, but we don't offer a way of checking for positive numbers including zero or negative numbers including zero. There is currently no way to check explicity for zero, therefore it isn't possible to do `either(isZero, isPositive)`. 

I suggest adding an `isZero` which would then allow us to add `isPositiveIncludingZero` and `isNegativeIncludingZero`.

### PR DESCRIPTION
**Pull request by Undistraction**
Add predicates for testing numbers for being non-positive (zero and all negative numbers) or non-negative (zero and all positive numbers)

Closes #324

### CODE REVIEW

File: src/index.d.ts [hunk 1]
```diff
@@ -251,15 +251,29 @@ declare namespace RamdaAdjunct {
         isNotNumber(val: any): boolean;
 
         /**
-         * Checks if value is a positive `Number` primitive or object.
+         * Checks if value is a positive `Number` primitive or object. Zero is not considered
+         * negative.
```
Comment (char0n) [inline src/index.d.ts]: Checks if value is a positive `Number` primitive or object. Zero is considered neither positive nor negative.

File: src/index.d.ts [hunk 2]
```diff
@@ -251,15 +251,29 @@ declare namespace RamdaAdjunct {
...
          */
         isPositive(val: any): val is number;
 
         /**
-         * Checks if value is a negative `Number` primitive or object.
+         * Checks if value is a negative `Number` primitive or object. Zero is not considered
```
Comment (char0n) [inline src/index.d.ts]: Checks if value is a negative `Number` primitive or object. Zero is considered neither positive nor negative.

File: src/index.d.ts [hunk 3]
```diff
@@ -251,15 +251,29 @@ declare namespace RamdaAdjunct {
...
         isNegative(val: any): val is number;
 
+        /**
+         * Checks if value is a non-positive `Number` primitive or object. This includes all
+         * negative numbers and zero.
+         */
+        isNonPositive(val: any): val is number;
+
+        /**  
+         * Checks if value is a non-positive `Number` primitive or object. This includes all
```
Comment (char0n) [inline src/index.d.ts]: Checks if value is a non-negative `Number` primitive or object. This includes all positive numbers and zero.

File: src/index.d.ts [hunk 4]
```diff
@@ -754,16 +768,16 @@ declare namespace RamdaAdjunct {
          * from fromIndex (inclusive).
          * Dispatches to the slice method of the third argument, if present.
          */
-        sliceFrom<T>(fromIndex: number, list: string|T[]): string|T[];
-        sliceFrom(fromIndex: number): <T>(list: string|T[]) => string|T[];
+        sliceFrom<T>(fromIndex: number, list: string | T[]): string | T[];
```
Comment (char0n) [inline src/index.d.ts]: Just for info: All these changes in typings are related to new version of tslint ?

File: src/isNegative.js [hunk 5]
```diff
@@ -3,7 +3,7 @@ import { both, gt } from 'ramda';
 import isNumber from './isNumber';
 
 /**
- * Checks if value is a negative `Number` primitive or object.
+ * Checks if value is a negative `Number` primitive or object. Zero is not considered negative.
```
Comment (char0n) [inline src/isNegative.js]: Checks if value is a negative `Number` primitive or object. Zero is considered neither positive nor negative.

File: src/isPositive.js [hunk 6]
```diff
@@ -3,7 +3,7 @@ import { both, lt } from 'ramda';
 import isNumber from './isNumber';
 
 /**
- * Checks if value is a positive `Number` primitive or object.
+ * Checks if value is a positive `Number` primitive or object. Zero is not considered positive.
```
Comment (char0n) [inline src/isPositive.js]: Checks if value is a positive `Number` primitive or object. Zero is considered neither positive nor negative.

File: test/isNonPositive.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
+import * as RA from '../src/index';
+import MAX_SAFE_INTEGER from '../src/internal/polyfills/Number.MAX_SAFE_INTEGER';
+import MIN_SAFE_INTEGER from '../src/internal/polyfills/Number.MIN_SAFE_INTEGER';
+import eq from './shared/eq';
+import args from './shared/arguments';
+import Symbol from './shared/Symbol';
+
+describe('isNonPositive', function() {
+  it('tests a value for positive `Number`', function() {
```
Comment (char0n) [inline test/isNonPositive.js]: tests a value for non-positive `Number`'

File: src/isNonNegative.js [hunk 8]
```diff
@@ -0,0 +1,30 @@
+import { both, gte, flip } from 'ramda';
+
+import isNumber from './isNumber';
+
+/**
+ * Checks if value is a non-positive `Number` primitive or object. This includes all positive
```
Comment (char0n) [inline src/isNonNegative.js]: Checks if value is a non-negative `Number` primitive or object. This includes all positive numbers and zero.

File: src/isNonNegative.js [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+ * numbers and zero.
+ *
+ * @func isNonNegative
+ * @memberOf RA
+ * @since {@link https://char0n.github.io/ramda-adjunct/1.15.0|v1.15.0}
```
Comment (char0n) [inline src/isNonNegative.js]: `@since {@link https://char0n.github.io/ramda-adjunct/2.6.0|v2.6.0}`

File: src/isNonNegative.js [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+ * @category Type
+ * @sig * -> Boolean
+ * @param {*} val The value to test
+ * @return {boolean}
+ * @see {@link RA.isPositive|isPositive}
```
Comment (char0n) [inline src/isNonNegative.js]: `@see {@link RA.isPositive|isPositive}, {@link RA.isNonPositive|isNonPositive}`

Also I would add the following into the `isNegative` source:
`@see {@link RA.isNonNegative|isNonNegative}`

File: src/isNonNegative.js [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+ * RA.isNonNegative(0); // => true
+ * RA.isNonNegative(1); // => true
+ * RA.isNonNegative(Infinity); // => true
+ * RA.isNonNegative(Number.MAX_VALUE); // => true
+ *
+ * RA.isNonNegative(-Infinity); // => false
+ * RA.isNonNegative(Number.MIN_VALUE); // => false
+ * RA.isNonNegative(NaN); // => false
+ */
+const isNonNegative = both(isNumber, flip(gte)(0));
```
Comment (char0n) [inline src/isNonNegative.js]: Would this be better for composition ?
```js
const isNonNegative = both(isNumber, complement(isNegative));
```

File: src/isNonPositive.js [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+
+import isNumber from './isNumber';
+
+/**
+ * Checks if value is a non-positive `Number` primitive or object. This includes all negative
+ * numbers and zero.
+ *
+ * @func isNonPositive
+ * @memberOf RA
+ * @since {@link https://char0n.github.io/ramda-adjunct/1.15.0|v1.15.0}
```
Comment (char0n) [inline src/isNonPositive.js]: `@since {@link https://char0n.github.io/ramda-adjunct/2.6.0|v2.6.0}`

File: src/isNonPositive.js [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+ * @category Type
+ * @sig * -> Boolean
+ * @param {*} val The value to test
+ * @return {Boolean}
+ * @see {@link RA.isNegative|isNegative}
```
Comment (char0n) [inline src/isNonPositive.js]: `@see {@link RA.isNegative|isNegative}, {@link RA.isNonNegative|isNonNegative}`

Also I would add the following into the `isPositive` source:
`@see {@link RA.isNonPositive|isNonPositive}`

File: src/isNonPositive.js [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+ * RA.isNonPositive(0); // => true
+ * RA.isNonPositive(-1); // => true
+ * RA.isNonPositive(-Infinity); // => true
+ * RA.isNonPositive(Number.MIN_VALUE); // => true
+ *
+ * RA.isNonPositive(Infinity); // => false
+ * RA.isNonPositive(Number.MAX_VALUE); // => false
+ * RA.isNonPositive(NaN); // => false
+ */
+const isNonPositive = both(isNumber, flip(lte)(0));
```
Comment (char0n) [inline src/isNonPositive.js]: Would this be better for composition ?
```js
const isNonPositive = both(isNumber, complement(isPositive));
```

Comment (Undistraction) [reply src/isNonPositive.js]: Does that make sense though given that `isPositive` is also checking for `isNumber`? I was thinking that this might complicate things, but I guess we're rechecking `isNumber` anyway. However this would mean we are checking it twice.

Comment (char0n) [reply src/isNonPositive.js]: I always prefer composition even though It may not be the most effective thing how to do things. We can always optimize later. I consider checking `isNumber` in `isPositive` an implementation detail in this case.

Comment (Undistraction) [reply src/isNonPositive.js]: OK. I'll buy that for a dollar.

Comment (Undistraction) [reply src/isNonPositive.js]: @char0n My original intuition was correct. The problem comes with `NaN`:

```
const isNonPositive = both(isNumber, complement(isPositive));
isNonPositive(NaN) // true
```

`NaN` is neither positive or non-positive, so this should be `false` shouldn't it? Using `complement` we are effectively doing `!both(isNumber, isPositive)`,  so NaN passes `isNumber` but fails `isPositive` giving a `false`, however the complement is `true`.

File: src/index.d.ts (hunk 4 above)
Comment (char0n) [reply src/index.d.ts]: Yes could be part of this PR. The question was rather if  the tslint rules changes to create these changes or did you do them manually ?

File: src/isNonPositive.js (hunk 14 above)
Comment (char0n) [reply src/isNonPositive.js]: Yeah, so the question is - should we consider NaN a non-positive number ? IMHO, I think no (but that is debatable). By this assumption we call use your original signature that seems to avoid NaN problem. But I would add note to function description that NaN is not considered non-positive number.

File: src/index.d.ts (hunk 4 above)
Comment (Undistraction) [reply src/index.d.ts]: No. I haven't made any changes outside of the immediate scope of the PR.

**Comment by Undistraction** [comment]
@char0n Just to summarise the outstanding changes: 

- Refactor isNonPositive to use `complement` of `isPositive` - Problem with `NaN` (see comment above).
- Refactor isNonNagative to use `complement` of `isPositive` - Problem with `NaN` (see comment above).
- sliceFrom typescript defs - Weren't changed as part of this PR.

**Comment by char0n** [comment]
@Undistraction thanks for summary, fallback to your original composition and let me know when you want final review.

File: src/isNonPositive.js (hunk 14 above)
Comment (Undistraction) [reply src/isNonPositive.js]: I think this is the most practical solution.

**Comment by Undistraction** [comment]
@char0n Should be good to go.
